File: kh2_src/tabs/inventory.py
from dash import Dash, html, dcc, callback, Input, Output, State, ALL
import kh2_src.kh2_utils as utils


@callback(
    Output("InventoryDiv", "children"),
    Input("InventoryTabs", "value"),
)
def __create_inventory(tab):
    kh2 = utils.kh2
    inventory = html.Div([
        html.Div([
            html.Label(k + ": "),
            dcc.Input(
                id={"type": "Inventory", "index": kh2.inventory_dict[k]},
                type="number",
                value=kh2.inventory[kh2.inventory_dict[k]],
                min=0,
                max=99,
                step=1,
                style={"width": 50},
            ),
        ],
            style={"margin-top": 20, "gap": 10},
        ) for k in kh2.stock_dict[tab] if kh2.inventory_dict[k] < len(kh2.inventory)
    ])
    return inventory

def create_inventory():
    kh2 = utils.kh2
    itabs = dcc.Tabs(id="InventoryTabs", value="Consumables")
    itabs.children = [
        dcc.Tab(label=k, value=k) for k in kh2.stock_dict
    ]
    return html.Div([
        itabs,
        html.Div(id="InventoryDiv"),
    ])

@callback(
    Input({"type": "Inventory", "index": ALL}, "value"),
    State({"type": "Inventory", "index": ALL}, "id"),
)
def inventory_callback(values, ids):
    kh2 = utils.kh2
    try:
        i = 0
        for id in ids:
            idx = id["index"]
            kh2.inventory[idx] = values[i]
            i += 1
    except:
        pass

File: kh2_src/tabs/inventory.js
export function create_inventory() {
    const kh2div = document.getElementById("kh2div");
    kh2div.innerHTML = `
    <div>
        <div>
            <h3>Category:</h3>
            <select id="kh2invtabs">
                ${
                    Object.keys(window.kh2.stock_dict)
                    .map((k) => `<option value="${k}">${k}</option>`)
                    .join("")
                }
            </select>
        </div>
        <div id="kh2invdiv"></div>
    </div>`;
    const kh2invtabs = document.getElementById("kh2invtabs");
    function tab_sel() {
        __create_inventory(kh2invtabs.value);
    }
    kh2invtabs.addEventListener("change", tab_sel);
    __create_inventory("Consumables");
}

function __create_inventory(tab) {
    const kh2invdiv = document.getElementById("kh2invdiv");
    const items = window.kh2.stock_dict[tab]
        .filter((label) => window.kh2.inventory_dict[label] < window.kh2.inventory.length);
    const inventory = items
        .map((label) => `
        <label style="display: flex; gap: 10px; align-items: center">
            ${label}: 
            <input
                type="number"
                min=0
                max=99
                step=1
                value=${window.kh2.inventory[window.kh2.inventory_dict[label]]}
                name=${window.kh2.inventory_dict[label]}
            >
        </label>`)
        .join("");
    kh2invdiv.innerHTML = `
    <div id="inventory">
        <h3>Inventory:</h3>
        ${inventory}
    </div>`;
    inventory_callbacks();
}

function inventory_callbacks() {
    const inventory = document.getElementById("inventory");
    inventory.addEventListener("change", (e) => {
        if (e.target.validity.valid)
            window.kh2.inventory[e.target.name] = e.target.value;
        e.target.value = window.kh2.inventory[e.target.name];
    });
}
